pkg/endpoint/policy.go

```diff
@@ -457,11 +526,16 @@ func (e *Endpoint) regenerate(owner Owner) error {
 	if e.Consumable != nil {
 		// Regenerate policy and apply any options resulting in the
 		// policy change.
-		if _, err := e.regeneratePolicy(owner); err != nil {
+		_, consumersToRm, err := e.regeneratePolicy(owner)
+		if err != nil {
 			e.Mutex.Unlock()
 			return fmt.Errorf("Unable to regenerate policy for '%s': %s",
 				e.PolicyMap.String(), err)
 		}
+
+		if len(consumersToRm) != 0 {
+			go owner.CleanCTEntries(e, consumersToRm)
```

Comment:
I wonder what happens if we get into an adverse condition where policies are changed often, how this will handle? I could imagine this ending up with several goroutines waiting to grab the endpoint lock and blocking everyone else out for extended periods as they iterate through all of the underlying BPF maps and clean up the CT entries.

Should we limit the number of queued goroutines for this and other similar cleanup calls?

Reply:
@joestringer The endpoint lock is only needed to get the IPs of the endpoint, and in case of a failure to log that failure in the endpoints' log. I've change the CleanCTEntries, to receive the endpoint's IPs, does that help?

Reply:
@aanm The thing that I'm concerned about is something like:

* There are lots of connections flowing through the node, meaning that we have lots of CT entries. To iterate through these kernel entries as part of `RunGC()`, it will take a long time.
* Policy is changing often, perhaps legitimately, or perhaps because of some other bug - for instance, #1886.
* In this kind of corner, stress-test case, we could end up queueing a large number of goroutines that each grabs the endpoint lock in `PolicyDelete()`, then proceed to perform an expensive map dump / sweep from the kernel CT map while holding the lock. The next one will block until it gets a chance at the lock. Should we limit the number of goroutines that may be started that end up waiting on this lock?

Reply:
@joestringer Can you verify if your concern is still valid? The GC will no longer use the endpoint's lock to clean the entries.

Reply:
Right, it looks like they won't lock out others like, for example, commandline cilium clients.

I think the only danger now is that such a situation could end up maxing out CPU as multiple threads dump through all of the ctmap entries in parallel.